pyface/action/schema/action_manager_builder.py

```diff
@@ -0,0 +1,277 @@
+
+    def create_action_manager(self, schema):
+        """ Create a manager for the given schema using the additions.
+
+        Any additions whose paths do not match the supplied
+
+        Parameters
+        ----------
+        schema : Schema
+            An Schema for an ActionManager subclass (ie. one of MenuBarSchema,
+            MenuSchema, or ToolBarSchema).
+
+        Returns
+        -------
+        manager : ActionManager
+            The concrete ActionManager instance corresponding to the Schema
+            with addtions.  This does not yet have concrete toolkit widgets
+            associated with it: usually those will be created separately.
+        """
+        additions_map = defaultdict(list)
+        for addition in self.additions:
+            if addition.path:
+                additions_map[addition.path].append(addition)
+
+        manager = self._create_action_manager_recurse(schema, additions_map)
+        manager.controller = self.controller
+        return manager
+
+    def prepare_item(self, item, path):
+        """ Called immediately after a concrete Pyface item has been created
+        (or, in the case of items that are not produced from schemas,
+        immediately before they are processed).
+
+        This hook can be used to perform last-minute transformations or
+        configuration. Returns a concrete Pyface item.
+
+        Parameters
+        ----------
+        item : pyface.action item
+            A concrete Pyface item.
+        path : str
+            The path to the item in the Schema.
+
+        Returns
+        -------
+        item : pyface.action item
+            A modified or transformed Pyface item.
+        """
+        return item
+
+    # ------------------------------------------------------------------------
+    # Private interface.
+    # ------------------------------------------------------------------------
+
+    def _get_ordered_schemas(self, schemas):
+        begin = []
+        middle = []
+        end = []
+
+        for schema in schemas:
+            absolute_position = getattr(schema, "absolute_position", None)
+            if absolute_position is None:
+                middle.append(schema)
+            elif absolute_position == "last":
+                end.append(schema)
+            else:
+                begin.append(schema)
+
+        schemas = (
+            before_after_sort(begin)
+            + before_after_sort(middle)
+            + before_after_sort(end)
+        )
+        return schemas
+
+    def _group_items_by_id(self, items):
+        """ Group a list of action items by their ID.
+
+        Action items are Schemas and Groups, MenuManagers, etc.
+
+        Return a dictionary {item_id: list_of_items}, and a list containing
+        all the ids ordered by their appearance in the `all_items` list. The
+        ordered IDs are used as a replacement for an ordered dictionary, to
+        keep compatibility with Python <2.7 .
```

Comment:
not relevant to this PR, but doing special things to keep py 2.7 compatibility could be removed / refactored at some point

Reply:
Yes,I noticed this and decided not to mess with it in this PR.